Common subdirectories: chdkptp/chdk_headers and chdkptp-upt/chdk_headers
diff -u chdkptp/chdkptp.c chdkptp-upt/chdkptp.c
--- chdkptp/chdkptp.c	2021-06-12 20:58:32.628701000 -0400
+++ chdkptp-upt/chdkptp.c	2021-09-10 15:33:22.956401185 -0400
@@ -2658,7 +2658,13 @@
 };
 
 
-static int chdkptp_registerlibs(lua_State *L) {
+int luaopen_chdkptp(lua_State *L) {
+
+	usb_init();
+	luaopen_lfs(L);
+	luaopen_lbuf(L);
+	luaopen_rawimg(L);	
+
 	/* set up meta table for error object */
 	luaL_newmetatable(L,CHDK_API_ERROR_META);
 	lua_pushcfunction(L,api_error_tostring);
@@ -2728,13 +2734,9 @@
 	/* register signal handlers */
 //	signal(SIGINT, ptpcam_siginthandler);
 	init_timing();
-	usb_init();
 	lua_State *L = luaL_newstate();
 	luaL_openlibs(L);
-	luaopen_lfs(L);
-	luaopen_lbuf(L);
-	luaopen_rawimg(L);
-	chdkptp_registerlibs(L);
+	luaopen_chdkptp(L);
 	int r=exec_lua_string(L,"require('main')");
 	uninit_gui_libs(L);
 	lua_close(L);
Only in chdkptp-upt/: chdkptp.so
Only in chdkptp-upt/: config.mk
Only in chdkptp-upt/: .dep
Only in chdkptp-upt/: extlibs
diff -u chdkptp/include.mk chdkptp-upt/include.mk
--- chdkptp/include.mk	2021-09-07 18:02:43.911264000 -0400
+++ chdkptp-upt/include.mk	2021-09-10 15:33:22.956401185 -0400
@@ -32,7 +32,7 @@
 EXE_EXTRA=
 
 CC=gcc
-CFLAGS=-DCHDKPTP_OSTYPE=\"$(OSTYPE)\" -Wall
+CFLAGS=-DCHDKPTP_OSTYPE=\"$(OSTYPE)\" -Wall -fPIC
 LDFLAGS=
 #LD=ld
 
Common subdirectories: chdkptp/lfs and chdkptp-upt/lfs
Common subdirectories: chdkptp/lua and chdkptp-upt/lua
Common subdirectories: chdkptp/lua-signal and chdkptp-upt/lua-signal
diff -u chdkptp/Makefile chdkptp-upt/Makefile
--- chdkptp/Makefile	2020-11-23 00:34:19.820041000 -0500
+++ chdkptp-upt/Makefile	2021-09-10 15:38:26.022798495 -0400
@@ -139,14 +139,19 @@
 endif
 
 CHDKPTP_EXE=chdkptp$(EXE_EXTRA)$(EXE)
+CHDKPTP_LIB=chdkptp$(EXE_EXTRA)$(DLL)
 
 EXES=$(CHDKPTP_EXE)
+LIBS=$(CHDKPTP_LIB)
 
-all: $(EXES)
+all: $(EXES) $(LIBS)
 
 SRCS=properties.c ptp.c chdkptp.c lbuf.c liveimg.c rawimg.c luautil.c $(PTPIP_SRCS)
 OBJS=$(SRCS:.c=.o)
 
+$(CHDKPTP_LIB): $(OBJS)
+	$(CC) -shared -o $@ lfs/lfs.o $^ $(LDFLAGS)
+
 $(CHDKPTP_EXE): $(OBJS)
 	$(CC) -o $@ lfs/lfs.o $^ $(LDFLAGS)
 
Common subdirectories: chdkptp/misc and chdkptp-upt/misc
